Add initial get_url()/open() api implementation

Suor opened the pull request:
No streaming so far. Everything is added to `dvc.api`, I can move to somewhere else something if wring with that.

Also a part of this PR:
- simplified google creds handling.
- tests parametrized by remotes - if approach is accepted I will refactor the rest of the tests in `test_data_cloud.py`
- added `Repo.get_remote()` as a first step to dropping `DataCloud`.
- fixed `PathHDFS` stringification.

efiop commented on dvc/api.py:
```diff
@@ -0,0 +1,25 @@
+    """Returns an url of `path` in default or specified remote"""
+    repo = Repo(repo_dir)
+    out, = repo.find_outs_by_path(path)
+    # import ipdb; ipdb.set_trace()
```
```suggestion
```

efiop commented on dvc/api.py:
```diff
@@ -0,0 +1,25 @@
+
+def get_url(path, repo_dir=None, remote=None):
+    """Returns an url of `path` in default or specified remote"""
+    repo = Repo(repo_dir)
```
Why not just make get_url part of the Repo? What is the point of having it separately, if you initialize repo anyway?

Suor replied:
To not expose `Repo` as public api.

efiop replied:
But Repo _is_ a public api. Similar to Repo() from gitpython.

Suor replied:
As far as I can see our public API is command line only. No internals are documented. And anyway this is the way we discussed it with @shcheklein and @dmpetrov - make it separately, so that users won't need to know about `Repo` and everything.

shcheklein replied:
just to clarify - does it accept URL as a repo_dir? Will Repo be responsible for running git clone?

There was a discussion indeed - all 4 of us were involved and we agreed to not expose Repo yet. It also feels that this API is not the same as the Repo one. It's like gitpython vs requests or something. One to deal with DVC CLI on a python level, another one is a higher level API for Python. The difference could be a context manager for an object, or a reading file into memory vs to the disk (the way we do in Repo as I would assume).

Suor replied:
It doesn't yet support url as repo dir. From what @efiop told me I thought it should be done on top of packages.

efiop commented on dvc/api.py:
```diff
@@ -0,0 +1,25 @@
+
+def open(path, repo_dir=None, remote=None):
+    """Opens a specified resource as a file descriptor"""
+    repo = Repo(repo_dir)
```
Same here. Something like
```
repo = Repo(repo_dir)
with repo.open(path) as fobj:
   ...
```
Don't really see a point of making it a separate `dvc/api`, if `Repo` is the API anyways.

Suor replied:
This was part of initial discussion and we decided to make this separate.

efiop replied:
It feels like that logic belongs to Repo. Even if we need get_url detached from Repo for convenience, we could simply make it a wrapper around a more generic Repo API. Something like:
```
def open(path, repo_dir=None, remote=None):
    return Repo(repo_dir).open(path, remote=remote)
```
What do you think?

dmpetrov replied:
This api is going to be publicly released soon with this "model deployment" feature. I'm not sure we are ready to release `Repo` api at that time. So, it makes sense to keep the public part (`dvc.api`) and upcoming full api (`dvc.repo`) separately at least for now.

It might be a good idea to create wrappers in `Repo` if other modules can benefit from this. Otherwise, I'd suggest postponing these activities until we will be ready to polish and release `repo` api.

efiop replied:
@dmpetrov The approach I'm suggesting above doesn't actually mean that we need to make Repo officially public. I'm just saying that `open`'s logic is using Repo internals and in the current state it looks like it belongs to Repo. The "api" part that we want to release is simply going to use Repo's public methods inside and act as a wrapper

Suor replied:
I am with @dmpetrov here, create methods in `Repo` when it needed later.

dmpetrov replied:
One more thought about the Repo API. Today it is not clear what kind of dvc-API need to be released. When we go through use cases we might decide that Repo is not what users need.

Why?
1. Repo class reflects CLI\procedural style where a more functional or OOP style might be a better fit for programmatic API of DVC. For example, AirFlow uses OOP style when the major object is step\stage, not Repo state.
2. Public API will require a lot of syntax sugar that will explode the size of Repo (which is already large class with a number of code smells).
3. We might decide to focus on python functions (not just file\commands) as a basic pipeline entity which is a much higher level of abstraction compared to Repo.

So, I'd not overfocus on external Repo API for now. We should focus on internal code design for developers(for ourselves).

efiop commented on dvc/api.py:
```diff
@@ -0,0 +1,25 @@
+        raise ValueError("Can't open a dir")
+
+    with repo.state:
+        cache_info = repo._collect_used_cache(out, remote=remote)
```
Using private method is not nice.

efiop commented on dvc/api.py:
```diff
@@ -0,0 +1,25 @@
+        repo.cache.local.pull(cache_info, remote=repo.get_remote(remote))
+
+    cache_filename = repo.cache.local.checksum_to_path(out.checksum)
+    return _open(cache_filename)
```
Maybe some mode flags, no?

Suor replied:
Maybe)

efiop commented on dvc/repo/__init__.py:
```diff
@@ -538,3 +539,29 @@ def func(out):
         path_parts = os.path.normpath(path).split(os.path.sep)
         return self.DVC_DIR in path_parts
+
+    def get_remote(self, remote=None):
```
So do we need DataCloud any longer?

Suor replied:
Need to reimplement `push`/`pull` etc without it first.

efiop replied:
I suppose get_remote is just an incomplete copypaste of DataCloud._get_cloud?

Suor replied:
I left compat thing there, no point refactoring future dead code.

efiop replied:
Maybe you could use get_remote in DataCloud, so there is no copypasting?

shcheklein replied:
Looks like repo file/class has gotten too large (number of function per file, number of lines, etc)- check CC issues. 

Can create a ticket and PR to simplify discussion for this RFC? To keep this one simple and merge faster?

Unless I'm missing something and it's actually needed for the implementation.

efiop replied:
Actually, most of the "get_remote" can be a part of config, since most of it is just prioritizing things from config one over each other. I think this should be solved before merging, because having copypasted stuff from DataCloud here is not good.

Suor replied:
I don't think `.get_remote()` is a good part of config. This way we should move all instantiation to config making it a babylonian tower.

Suor replied:
I can continue removing DataCloud in this PR.

shcheklein replied:
Would be great to see some motivation behind this decision and some description of the suggested improvement. It's hard to review it w/o seeing where is it going.

Also, like I mentioned repo is already too complicated: `> 400 lines` + `> 20 methods`. There are Code Climate complaints about this change.

efiop commented on tests/func/test_data_cloud.py:
```diff
@@ -164,7 +164,7 @@ def get_local_url():
 
 def get_ssh_url():
-    return "ssh://{}@127.0.0.1:{}".format(
+    return "ssh://{}@127.0.0.1:22{}".format(
```
Is this really needed? 22 should be used by default already.

Suor replied:
This is needed as long as we compare urls as strings. It was easier for me to change it here, than change a test.

Ultimately, we need a proper path refactor, which will solve all such inconsistencies. I will try to add it in a separate PR tomorrow.

efiop commented on tests/func/test_data_cloud.py:
```diff
@@ -894,3 +894,79 @@ def test(self):
+
+
+def test_api_open(repo_dir, dvc_repo, remote):
+    main(["remote", "add", "-d", "upstream", remote.get_url()])
```
Main returns an integer, which you should check. Otherwise, you might be missing an error.

efiop replied:
I have config.remote_add coming in my pkgs patch, which should help us avoid using CLI for config modifications. So it would be something like `dvc_repo.config.remote_add`, so exceptions won't be caught like CLI does.

Suor replied:
As I said above I can't see complicating config with outer knowledge as a good thing. If I would add a way to modify config I would make it in a way that it won't know about its contents. So:

```python
remote_name = "..."
remote_section = Config.REMOTE_SECTION_FMT.format(remote_name)
config[remote_section] = {...}
config["core"]["remote"] = remote_name
```

somewhere in remote or repo - the same place as we do `.get_remote()`.

efiop commented on tests/func/test_data_cloud.py:
```diff
@@ -894,3 +894,79 @@ def test(self):
+    )
+
+
+def test_api_open(repo_dir, dvc_repo, remote):
```
If you are testing "api", maybe just create a new test file instead of appending all of this to test_data_cloud? Might even consider adding a directory like `tests/func/remote/`.

Suor replied:
As long as I use all those `_should_test()` and `_get_url()` it's far easier to have it here. Once `DataCloud` is gone this could be separated into `test_remotes.py` and `test_api.py`.

Suor replied:
I am just being incremental here.

efiop replied:
@Suor Nothing prevents you from using _should_test and _get_url in a test_api.py. No need to make test_data_cloud.py even worse than it is already :slightly_smiling_face:

efiop commented on dvc/api.py:
```diff
@@ -0,0 +1,24 @@
+    if out.isdir():
+        raise ValueError("Can't open a dir")
+
+    with repo.state:
```
Here API is really using Repo internals(e.g. you've renamed collect_used_cache to make it public, but it shouldn't be exposed actually), I would rather move this into `dvc/repo/open.py` and have
```
def open(path, repo_dir=None, remote=None):
    return Repo(repo_dir).open(path, remote=remote)
```
here.

Suor replied:
Not using internals anymore. And I don't want to create method that is called only once.

shcheklein replied:
I don't understand either why are we exposing some internal utility method here. Repo _is_ API even if we decided to keep it hidden for now. It was made to be an API and we try to keep is clean. If there is a need for a simple mechanism to reuse certain helper, let's introduce _internal_ package or something.

Suor commented:
Latest changes:
- moved `.get_remote()` to `DataCloud` to replace `._get_cloud()` and `_cloud`
- moved `.open()` implementation inside `Repo`
- moved tests
- added `open()` flags `mode` and `encoding` as most required and ones we can implement for streaming.

Removing `DataCloud` or simplifying `Repo` is a refactoring to big to perform it here. I will create a separate issue to track it.
